Issue 95 solution

jcyannotty opened the pull request:
Starting the PR for the Trees update. All tests were passing at completion of this update - one minor change that might need to be made is with the import step with OpenBT (particularly in the tests).

jared321 commented:
@jcyannotty Is it OK if I change `from openbtmixing import OpenbtMix` to `from openbtmixing import Openbtmix` to match the correct case in the openbtmixing package?

When I run tests on this branch using the openbtmixing package installed from `6_MesonProto`, `test_mixing` fails at both of the following tests
```
assert fit["pbd"] == 0.7, "check prob of birth and death"
assert fit["pb"] == 0.5, "check prob of birth"
```
If I comment these two out, then all tests are passing.  I see that `fit` is the dict returned by the `train` function.  On lines 334-338 in `openbtmixing.py` in `main`, it appears that this dict is being created with only the following keys
```
res = {}
keys = ['fpath', 'inform_prior', 'minnumbot', 'nummodels', 'p']

for key in keys:
    res[key] = self.__dict__[key]
```
I suspect that previous testing was passing here if they were using openbtmixing installed from the old binary wheels available at PyPI.  This assumes that those key/value pairs were once included in the training result.

How do you want me to fix this?  Remove those two tests?  Add `pb` and `pbd` to `keys`?

jared321 commented:
As per the git workflow, `main` needs to be merged into your branch so that it is as closely aligned to `main` as possible.  This will also force addressing the current `trees.py` merge conflict as a merge commit on your branch rather than on `main`.  Unfortunately, there are a lot of changes on `trees.py`, so I cannot perform that merge conflict resolution for you.  Please let me know how I can help with it.

Once that merge is completed, I propose the following
* I will update the GitHub actions so that they install openbtmixing from `6_MesonProto` rather than from PyPI
* If all tests are passing then
  * I will merge `6_MesonProto` into `main`
  * You will update PyPI so that openbtmixing is installed only from source distributions using meson
  * I will update GitHub actions again to install from PyPI.
* If all tests are still passing, then we can merge this PR once you and @asemposki are ready.

How does this sound?

jcyannotty commented:
I have resolved the merge conflicts. In response to previous comments, I have also modified the import step to read `from openbtmixing import Openbtmix` and removed the outdated assertions that were causing the tests to fail.

jared321 commented:
@jcyannotty Fantastic.  My tests are all passing after your fixes.

I don't see the merge of `main` into this branch but do still see in this PR that the merge is blocked due to conflicts on `trees.py`.  Can you please try doing that merge again and pushing the result?  If you would like, you can do that merge by creating a PR from `main` into your branch so that we have a record of that task in GH.  Let me know if you want help with this step.

I'll start work on the changes to the GH actions but won't commit until the conflicts are resolved on your end.

jared321 commented:
@asemposki Do you want these changes to be marked as a new Taweret version?  Are there other similar official release changes that will need to be made before this PR is merged into `main`?

jcyannotty commented:
I can take care of the "Code Standard Adherence" tests that are failing from the updated trees.py and test_trees.py files. Tests are failing, some having to do with the bivariate_linear module. This is unexpected, because this PR is only working with trees.

jared321 commented:
OK.  Let me fix up the GH actions to see if the other GH actions pass once they install OpenBTMixing using meson.

jared321 commented:
@jcyannotty @asemposki I've gotten the Taweret tests up and running using builds of OpenBTMixing from the `6_MesonProto` branch.  This includes modernizing the macOS versions that are tested and adding Python 3.13.  As in the OpenBTMixing repo, we are testing with both OpenMPI and MPICH where possible and have a GH action for testing with OpenMPI in anaconda.  So far, so good.  You both might want to take a look at the `run_*.yml` actions to see if you are happy with the setups that we are presently testing. 

I will fix the sphinx and Jupyter book actions if need be once the new openbtmixing package is in PyPI.

@jcyannotty I hand over the branch to you to get the standard checking action passing again.

@asemposki It looks like tests are failing with scipy >= v1.15.0.  Can you investigate this?
```
__________________ ERROR collecting test_bivariate_linear.py ___________________
ImportError while importing test module '/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/Taweret/tests/test_bivariate_linear.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/Taweret/tests/test_bivariate_linear.py:16: in <module>
    import bilby
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/bilby/__init__.py:21: in <module>
    from . import core, gw, hyper
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/bilby/core/__init__.py:1: in <module>
    from . import grid, likelihood, prior, result, sampler, series, utils, fisher
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/bilby/core/grid.py:6: in <module>
    from .prior import Prior, PriorDict
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/bilby/core/prior/__init__.py:1: in <module>
    from .analytical import *
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/bilby/core/prior/analytical.py:3: in <module>
    from scipy.special._ufuncs import xlogy, erf, log1p, stdtrit, gammaln, stdtr, \
E   ImportError: cannot import name 'btdtri' from 'scipy.special._ufuncs' (/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/scipy/special/_ufuncs.cpython-310-x86_64-linux-gnu.so)
```

asemposki commented:
I can investigate that; from cursory glance it looks like a problem with `bilby`, but I hope there's a reason outside of that package why this is occurring. 

In regard to the new Taweret version, I think this would be going from v1.1.0 to v.1.2.0, since we're doing substantial changes to the Trees module, but if it should be considered a more minor change, e.g., v1.1.0 to v.1.1.1, I'm open to discussion. As for other official release changes, I think the only other thing we would need to do is include information in the README, and potentially in the documentation, about the meson build (update the README for MacOS users). I cannot think of anything else at the moment that needs to be taken care of.

asemposki commented:
<img width="806" alt="Screenshot 2025-01-20 at 10 19 36" src="https://github.com/user-attachments/assets/5dec14ec-8a36-48c4-b8f9-257320c6ff02" />

From this commit, it looks like the most recent version of `scipy` eliminated that function (this was only 3 months ago), and only has now `btdtria` and `btdtrib` instead---they also removed the deprecation warning for this, so it is officially entirely removed. This version of scipy hence seems to be too far ahead of the one `bilby` developers have currently set up `bilby` to use---they need to update their use of this function then to comply with this version of `scipy`. I will keep poking around but it looks like we might not be able to use the latest `scipy` version with `bilby` until `bilby` fixes this issue.

jared321 commented:
@asemposki I have altered the anaconda GH action to get it passing again for all setups.  Refer to Issue #122 for more information.

jared321 commented:
@jcyannotty Something changed in the way the `build` package names wheels it creates.  This caused many of the actions to fail after your last commit.  I made some commits to fix that.

I still see the following PEP8 issues in the standards action:
```
./src/Taweret/mix/trees.py:12:1: E302 expected 2 blank lines, found 1
./src/Taweret/mix/trees.py:184:80: E501 line too long (83 > 79 characters)
./src/Taweret/mix/trees.py:217:1: W293 blank line contains whitespace
./src/Taweret/mix/trees.py:296:1: W293 blank line contains whitespace
./src/Taweret/mix/trees.py:343:1: W293 blank line contains whitespace
./src/Taweret/mix/trees.py:356:1: W293 blank line contains whitespace
./src/Taweret/tests/test_trees.py:65:33: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
./src/Taweret/tests/test_trees.py:65:80: E501 line too long (81 > 79 characters)
./src/Taweret/tests/test_trees.py:81:11: E251 unexpected spaces around keyword / parameter equals
./src/Taweret/tests/test_trees.py:81:13: E251 unexpected spaces around keyword / parameter equals
```

jcyannotty commented:
@jared321 The openbtmixing python package has been updated on pypi with the Meson build that you have implemented.

jared321 commented:
In progress work

- [x] Downloaded OpenBTMixing v1.1.0 source distribution and confirmed good contents
- [x] Built OpenBT virtual env with OpenBTMixing installed from PyPI and confirmed that tests run.
- [x] Built Taweret virtual env with Taweret installed by pip in editable mode from a local clone set to this branch.
   - [x] Confirmed that OpenBTMixing v1.1.0 built/installed correctly and see CLT binaries installed in venv
   - [x] Confirmed that Taweret test suite passes
- [ ] Run all tox tasks and confirm passing 
- [ ] Resetting/cleaning/modernizing GH actions so that they get OpenBTMixing from PyPI & build from source distribution
- [ ] Setting Taweret scipy version dependence as needed (Issue #123) & clean GH actions

jcyannotty commented:
> @jcyannotty Thanks for getting the source distribution uploaded. So far, most things are looking good. When your ["K" trees notebook](https://github.com/bandframework/Taweret/blob/main/book/notebooks/Trees_BMM/Trees_BMM_2D_K_models.ipynb) is run when building the Taweret Jupyterbook, I get the following error
> 
> ```
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> Cell In[9], line 6
>       3 mix = Trees(model_dict = model_dict)
>       5 # Set prior information
> ----> 6 mix.set_prior(k=2.5,ntree=10,overallnu=5,overallsd=0.01,inform_prior=False)
>       8 # Train the model
>       9 fit = mix.train(X=x_train, y=y_train, ndpost = 10000, nadapt = 2000, nskip = 2000, adaptevery = 500, minnumbot = 4, numcut = 100)
> 
> TypeError: Trees.set_prior() got an unexpected keyword argument 'overallnu'
> ```

Ah i renamed "overallnu" to "nu." I have updated the notebook and tests are running now. I'll monitor to see if anything else breaks

jared321 commented:
We're getting there!  Now I see a similar error for overallsd.  When I removed overallsd, it ran for a while but raised a different error
```
ValueError: Number of models in f_test does not equal number of models in f_train.
```
Not sure if that has to do with my removing overallsd or not.  I think that you need to make the same changes in your other notebook as well.
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 6
      3 mix = Trees(model_dict = model_dict)
      5 # Set prior information
----> 6 mix.set_prior(k=2.5,ntree=10,nu=5,overallsd=0.01,inform_prior=False)
      8 # Train the model
      9 fit = mix.train(X=x_train, y=y_train, ndpost = 10000, nadapt = 2000, nskip = 2000, adaptevery = 500, minnumbot = 4, numcut = 100)

TypeError: Trees.set_prior() got an unexpected keyword argument 'overallsd'
```

jared321 commented:
I see the following error for the K notebook
```
ValueError: Number of models in f_test does not equal number of models in f_train.
```
The other notebook fails at the very end with
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[19], line 2
      1 # Plot the posterior of the error standard deviation
----> 2 mix.plot_sigma()

AttributeError: 'Trees' object has no attribute 'plot_sigma'
```

jcyannotty commented:
Alright looks like we have a successful build! A couple things to note:
- I had to update openbtmixing to set the number of models within the train function. I had a default value of 2 so that is why things ran in the first notebook with 2 models but not in the second with K = 3 models. 
- I added a unit test in openbtmixing to check that the number of models is set based on the data passed into train() & also set the default to 1 rather than 2....that way we know the test will catch any issues.

I think we are ready to merge into main. Anything else we need to add? 

Lastly - thank you for your help throughout all of this!

jared321 commented:
That's fantastic!  I'm so happy that we got here and that your notebooks served as a way to ferret out a bug.  I've got some checkboxes to tick in a comment from above.  Then I'll hand things off to @asemposki  to make the final determination that we are ready for the next official Taweret release.

If you have time, could you download the Jupyterbook renderings artifact from here
https://github.com/bandframework/Taweret/actions/runs/14940561795
and confirm that you are happy with the content and renderings of your notebooks?

jcyannotty commented:
Looks like the artifact expired. I'll rerun the job and check back in 30 mins or so

jcyannotty commented:
Everything looks good on my end!

jared321 commented:
@jcyannotty Thanks for getting v1.1.2 in PyPI!  I updated all the actions so that they install that version automatically and all actions are passing again.

I just have a few quality checks to do still (see above) before I pass off to Alexandra for final acceptance.
